src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -299,6 +307,115 @@ static void AppendHashtableContents(IndentedTextWriter writer, Hashtable ht)
+                    writer.WriteLine("_timeout = (int)(timeout.TotalMilliseconds + 0.5); // Round;");
+                    writer.WriteLine("_timeoutOccursAt = global::System.Environment.TickCount + _timeout;");
+                    writer.WriteLine("_timeoutChecksToSkip = TimeoutCheckFrequency;");
+                }
+            }
+#pragma warning restore CS8321 // Local function is declared but never used
+
+            void EmitPrevLenCheck()
+            {
+                writer.WriteLine("// If previous match was empty or failed, advance by one before matching.");
+                using (EmitBlock(writer, "if (prevlen == 0)"))
+                {
+                    using (EmitBlock(writer, "if (textstart == text.Length)"))
+                    {
+                        writer.WriteLine("base.runmatch = global::System.Text.RegularExpressions.Match.Empty;");
+                        writer.WriteLine("return;");
+                    }
+                    writer.WriteLine();
+                    writer.WriteLine("base.runtextpos++;");
+                }
+            }
+
+            void EmitMainSearchLoop()
+            {
+                using (EmitBlock(writer, "while (true)"))
+                {
+                    using (EmitBlock(writer, "if (FindFirstChar(text))"))
+                    {
+                        writer.WriteLine("base.CheckTimeout();");
+                        writer.WriteLine();
+
+                        writer.WriteLine("// If we got a match, we're done.");
+                        using (EmitBlock(writer, "if (Go(text))"))
+                        {
+                            using (EmitBlock(writer, "if (quick)"))
+                            {
+                                writer.WriteLine("base.runmatch = null;");
+                                writer.WriteLine("return;");
+                            }
+                            writer.WriteLine();
+
+                            writer.WriteLine("// base.runmatch!.Tidy(base.runtextpos);");
+                            writer.WriteLine("return;");
+                        }
+                        writer.WriteLine();
+
+                        writer.WriteLine("// Reset state for another iteration.");
+                        writer.WriteLine("base.runtrackpos = base.runtrack!.Length;");
+                        writer.WriteLine("base.runstackpos = base.runstack!.Length;");
+                        writer.WriteLine("base.runcrawlpos = base.runcrawl!.Length;");
+                    }
+                    writer.WriteLine();
+
+                    writer.WriteLine("// We failed to find a match. If we're at the end of the input, then we are done.");
+                    using (EmitBlock(writer, "if (base.runtextpos == text.Length)"))
+                    {
+                        writer.WriteLine("base.runmatch = global::System.Text.RegularExpressions.Match.Empty;");
```

Comment:
Why is this needed?  If we initialize runmatch in the regex lib prior to calling Scan, and if Go ensures it uncaptures everything in the event of a failed match, can we just leave runmatch unmodified, and the regex lib will see that runmatch doesn't contain a top-level capture and will know that means a failed match?